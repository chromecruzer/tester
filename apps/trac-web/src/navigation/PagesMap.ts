import {BuildUrl} from "@trac/datatypes";
import _ from "lodash";

export type PageMapType = (
  'Accounts' |
  'Analytics' |
  'Details' |
  'Audit' |
  'Audited Accounts' |
  'Open Audits' |
  'Audit Upload' |
  'Audit Upload Session' |
  'Audit Work' |
  'Consignment' |
  'Customers' |
  'Data Upload' |
  'Expiration' |
  'Duplicates' |
  'Products' |
  'Products2' |
  'Login' |
  'SignUp' |
  'ChangePwd' |
  'Sales Reps' |
  'Search' |
  'Clear Search' |
  'Email Settings'
  );
export const pageMappings: { [index: string]: string } = {
  'Accounts': 'consignment/accounts',
  'Analytics': 'analytics/:upload_filetype/:session',
  'Details': 'details/:upload_filetype/:session',
  'Audit Work': 'audit/work',
  'Audited Accounts': 'audit/accounts',
  'Open Audits': 'audit/open',
  'Audit Upload': 'audit/upload/',
  'Audit Upload Session': 'audit/upload/:session',
  'Consignment': 'consignment',
  'Customers': 'data/customers',
  'Data Upload': 'data/upload',
  'Expiration': 'consignment/expiration',
  'Duplicates': 'data/duplicates',
  'Products': 'settings/products',
  'Products2': 'settings/products2',
  'Login': 'login',
  'SignUp': 'signup',
  'ChangePwd': 'changepwd',
  'Sales Reps': 'data/salesreps',
  'Search': '/search',
  'Clear Search': '/',
  'Email Settings': 'settings/email',
};
const dump = obj => JSON.stringify(obj, null, 2);

export default class PagesMap {
  private parentMap: { [index: string]: string } = {
    'audit': 'Audit',
    'consignment': 'Consignment',
    'data': 'Data',
    'settings': 'Settings',
    'orphan': 'Orphan',
  };
  private cleanUrlRegex: RegExp;

  constructor(private buildUrl = new BuildUrl()) {
    this.cleanUrlRegex = /[/]:.+$/;

  }

  page(page: PageMapType, paths: string[] = [], params: { [index: string]: (string | string[]) } = {}) {

    // console.log(`navigate to page ${page} = ${pageMappings[page]} transformed ${result}`);
    return this.buildUrl.build(this.clean(pageMappings[page]), paths, params);
  }

  private clean(url) {
    if (url) {
      return _.replace(url, this.cleanUrlRegex, '');
    }
    return url;
  }

  activeParent(url) {
    const tokens = _.filter(_.split(this.clean(url)), t => !_.isEmpty(t));
    let parent: (string | null) = null;
    if (tokens.length > 1) {
      parent = tokens[tokens.length - 2];
    }
    // console.log(`tokenize found a parent ${parent} from ${url}`);
    if (_.isNull(parent)) {
      return 'Orphan';
    }
    return this.parentMap[parent] || 'Orphan';
  }

  activePage(url) {
    if (url === '/') {
      return 'Search';
    }
    const result = _.find(_.keys(pageMappings), p => {//TODO: may have some ordering issues here
      if (p === 'Search') {
        return false;
      }
      return _.includes(url, this.clean(pageMappings[p]));
    }) || null;
    if (_.isNull(result)) {
      throw new Error(`URL ${url} does not correspond to a page`);
    }
    return result;
  }
}


