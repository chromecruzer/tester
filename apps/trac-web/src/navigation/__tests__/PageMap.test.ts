import PagesMap, {pageMappings} from "../PagesMap";

describe('PagesMap', () => {
  let pagesMap, mockBuildUrl;
  beforeEach(() => {
    mockBuildUrl = {
      build: jest.fn().mockReturnValue('this is a URL')
    };
    pagesMap = new PagesMap(mockBuildUrl);
  });
  describe('page', () => {
    it('should call buildURL with the right path', () => {
      pagesMap.page('Expiration');
      expect(mockBuildUrl.build.mock.calls).toMatchSnapshot();
    });
    it('should call buildURL with the right path parameters', () => {
      pagesMap.page('Details', ['audit-scan', 'SESSION']);
      expect(mockBuildUrl.build.mock.calls).toMatchSnapshot();
    });
    it('should call buildURL with the search parameters', () => {
      pagesMap.page('Search', [], {text: 'spencer', filters: ['customers']});
      expect(mockBuildUrl.build.mock.calls).toMatchSnapshot();
    });
  });
  describe('activePage', () => {
    it('should recognize all of the pages', () => {
      Object.keys(pageMappings).forEach(p => {
        // console.log(`testing url ${pageMappings[p]} for page ${p}`);
        const actual = pagesMap.activePage(pageMappings[p]);
        expect(actual).toEqual(p);
      })
    });
    it('should throw an error when it cannot figure out a URL', () => {
      expect(() => pagesMap.activePage('oopsie/daisy/bicycle')).toThrowErrorMatchingSnapshot();
    });
  });
  describe('activeParent', () => {
    it('should recognize all of the parents', () => {
      const testsCases = {
        'Snapshots': 'Orphan',
        'Search': 'Orphan',
        'Consignment': 'Consignment',
        'Expiration': 'Consignment',
        'Audit': 'Consignment',
        'Upload': 'Audit',
        'Analytics': 'Upload',
        'Details': 'Upload',
        'History': 'Snapshots',
        'Procedures': 'Orphan',
        'Accounts': 'Orphan',
        'Reports': 'Orphan',
        'Settings': 'Orphan'
      }
      Object.keys(PagesMap).forEach(p => {
        // console.log(`testing url ${PagesMap[p]} for page ${p}`);
        const actual = pagesMap.activeParent(PagesMap[p]);
        expect(actual).toEqual(testsCases[p]);
      })
    });
    it('should return orphan when it cannot figure out a URL', () => {
      const actual = pagesMap.activeParent('/oopsie/daisy/bicycle');
      expect(actual).toEqual('Orphan');
    });
  });
});
