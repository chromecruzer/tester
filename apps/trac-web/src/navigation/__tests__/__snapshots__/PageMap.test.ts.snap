// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagesMap activePage should throw an error when it cannot figure out a URL 1`] = `"URL oopsie/daisy/bicycle does not correspond to a page"`;

exports[`PagesMap page should call buildURL with the right path 1`] = `
Array [
  Array [
    "consignment/expiration",
    Array [],
    Object {},
  ],
]
`;

exports[`PagesMap page should call buildURL with the right path parameters 1`] = `
Array [
  Array [
    "details",
    Array [
      "audit-scan",
      "SESSION",
    ],
    Object {},
  ],
]
`;

exports[`PagesMap page should call buildURL with the search parameters 1`] = `
Array [
  Array [
    "/",
    Array [],
    Object {
      "filters": Array [
        "customers",
      ],
      "text": "spencer",
    },
  ],
]
`;
